### PR TITLE
Fix fault handling and mpu test address

### PR DESCRIPTION
**Pull request by kishore-sn**
- Fix fault handling: Restart only the app in which fault occured
- Fix mpu address:: Obtain kernel and app address from kernel ioctl

### CODE REVIEW

File: os/drivers/testcase/kernel_testcase_drv.c [hunk 1]
```diff
@@ -120,6 +133,31 @@ static int kernel_test_drv_ioctl(FAR struct file *filep, int cmd, unsigned long
 		break;
 #endif
 
+#ifdef CONFIG_EXAMPLES_MPU_TEST
+	case TESTIOC_MPUTEST:
+		ret = OK;
+		struct mputest_arg_s *obj = (struct mputest_arg_s*)arg;
```
Comment (jsdosa) [inline os/drivers/testcase/kernel_testcase_drv.c:139]: could you check null for arg?

Comment (jsdosa) [inline os/drivers/testcase/kernel_testcase_drv.c:139]: arg is unsigned long.
is it right?

Comment (kishore-sn) [reply os/drivers/testcase/kernel_testcase_drv.c:139]: Yes. Using arg like this is ok. It is common way of passing argument to ioctl.
I have added the null check.

File: os/drivers/testcase/kernel_testcase_drv.c [hunk 2]
```diff
@@ -120,6 +133,36 @@ static int kernel_test_drv_ioctl(FAR struct file *filep, int cmd, unsigned long
...
+		
+		if (!obj) {
+			return -EINVAL;
+		}
+
+		if (obj->type == MPUTEST_KERNEL_CODE) {
```
Comment (Taejun-Kwon) [inline os/drivers/testcase/kernel_testcase_drv.c:145]: For future work, it had better use switch-case instead of if-else.
If there's no change like that, please ignore this comment.

File: apps/examples/mpu/mpu_test.c [hunk 3]
```diff
@@ -86,14 +96,19 @@ int mpu_tc_main(int argc, char *argv[])
...
 
 				printf("ERROR: User Task made invalid access to Kernel code space\n");
 			} else if (ch == 'D' || ch == 'd') {
-			       address = (uint32_t *)CONFIG_MPUTEST_KERNEL_DATA_ADDR;
+				obj.type = MPUTEST_KERNEL_DATA;
+				ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
+				if (ret < 0) {
+					printf("ERROR: Failed to fetch test address from kernel\n");
+					return ERROR;
+				}
```
Comment (Taejun-Kwon) [inline apps/examples/mpu/mpu_test.c:111]: Blow is just example, you may understand what I mean,
```
if (ch == 'a' || ch == 'A') { // for input, put lowercase is always faster
    obj.type = MPUTEST_KERNEL_DATA;
} else if (ch == 'w' || ch == 'W) {
   oby.type = MPUTEST_APP_ADDR;
} else {
   ...
}

ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
close(tc_fd);
if (ret < 0) {
   printf("error : %d\n", errno); // ioctl in vfs layer will set errno.
   return ERROR;
}
return OK;
```

File: loadable_apps/loadable_sample/wifiapp/recovery.c [hunk 4]
```diff
@@ -44,18 +51,31 @@ static void *assert_thread(void *index)
...
+		ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
+		if (ret < 0) {
+			printf("ERROR: Failed to obtain address from kernel\n");
+			return ERROR;
+		}
+
+		printf("[%d] %dth thread, Write another app space 0x%x\n", getpid(), (int)index, obj.addr);
 		sleep(1);
-		*addr = 0xdeadbeef;
+		*(obj.addr) = 0xdeadbeef;
```
Comment (Taejun-Kwon) [inline loadable_apps/loadable_sample/wifiapp/recovery.c:75]: here also you can reduce duplicated code.
and 0xdeadbeef is only hexspeak value which we use for color?

File: apps/examples/mpu/mpu_test.c [hunk 5]
```diff
@@ -22,36 +22,43 @@
...
 
 #ifdef CONFIG_BUILD_KERNEL
 int main(int argc, FAR char *argv[])
 #else
 int mpu_tc_main(int argc, char *argv[])
 #endif
 {
-	volatile uint32_t *address;
 	char ch;
+	int tc_fd = open(KERNEL_TC_DRVPATH, O_WRONLY);
```
Comment (Taejun-Kwon) [inline apps/examples/mpu/mpu_test.c:36]: actually kernel_test_drv doesn't have open, how this works?????I can't belive it...

Comment (jeongchanKim) [inline apps/examples/mpu/mpu_test.c:36]: Taejun is right... There is no open in kernel_test_drv.
But this is not related to this commit. We should check/modify with another commit.

Anyway, originally `KERNEL_TC_DRVPATH` is for kernel tc. I think we'd better change the path name if you want to use in here, like `TESTCASE_DRVPATH` or something.

Comment (thapav) [inline apps/examples/mpu/mpu_test.c:36]: Kindly add declaration and definition separately.

File: apps/examples/mpu/mpu_test.c [hunk 6]
```diff
@@ -22,36 +22,43 @@
...
 		ch = getchar();
 
-#ifdef CONFIG_MPUTEST_APP_ADDR
 		if (ch == 'A' || ch == 'a') {
-			address = (uint32_t *)CONFIG_MPUTEST_APP_ADDR;
+			obj.type = MPUTEST_APP_ADDR;
+			ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
+			if (ret < 0) {
+				printf("ERROR: Failed to fetch test address from kernel\n");
+				return ERROR;
```
Comment (thapav) [inline apps/examples/mpu/mpu_test.c:60]: Please close tc_fd before return.

File: apps/examples/mpu/mpu_test.c [hunk 7]
```diff
@@ -63,21 +70,24 @@ int mpu_tc_main(int argc, char *argv[])
...
 			printf("\nPress C - For Read Code Test\n");
 			printf("Press D - For Read Data Test\n");
 
 			ch = getchar();
 			if (ch == 'C' || ch == 'c') {
-				address = (uint32_t *)CONFIG_MPUTEST_KERNEL_CODE_ADDR;
+				obj.type = MPUTEST_KERNEL_CODE;
+				ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
+				if (ret < 0) {
+					printf("ERROR: Failed to fetch test address from kernel\n");
```
Comment (thapav) [inline apps/examples/mpu/mpu_test.c:88]: Same as above, Please close tc_fd before return.

File: apps/examples/mpu/mpu_test.c [hunk 8]
```diff
@@ -86,14 +96,19 @@ int mpu_tc_main(int argc, char *argv[])
...
+				dest = *(obj.addr);
 
 				printf("ERROR: User Task made invalid access to Kernel code space\n");
 			} else if (ch == 'D' || ch == 'd') {
-			       address = (uint32_t *)CONFIG_MPUTEST_KERNEL_DATA_ADDR;
+				obj.type = MPUTEST_KERNEL_DATA;
+				ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
+				if (ret < 0) {
+					printf("ERROR: Failed to fetch test address from kernel\n");
+					return ERROR;
```
Comment (thapav) [inline apps/examples/mpu/mpu_test.c:110]: Same as above, Please close tc_fd before return.

File: apps/examples/mpu/mpu_test.c [hunk 9]
```diff
@@ -102,26 +117,28 @@ int mpu_tc_main(int argc, char *argv[])
...
 		ch = getchar();
 
-#ifdef CONFIG_MPUTEST_APP_ADDR
 		if (ch == 'A' || ch == 'a') {
-			address = (uint32_t *)CONFIG_MPUTEST_APP_ADDR;
+			obj.type = MPUTEST_APP_ADDR;
+			ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
+			if (ret < 0) {
+				printf("ERROR: Failed to fetch test address from kernel\n");
+				return ERROR;
```
Comment (thapav) [inline apps/examples/mpu/mpu_test.c:140]: Same as above, Please close tc_fd before return.

File: apps/examples/mpu/mpu_test.c [hunk 10]
```diff
@@ -132,53 +149,64 @@ int mpu_tc_main(int argc, char *argv[])
...
 			printf("Press D - For Write Kernel Data Test\n");
 
 			ch = getchar();
 			if (ch == 'C' || ch == 'c') {
-				address = (uint32_t *)CONFIG_MPUTEST_KERNEL_CODE_ADDR;
+				obj.type = MPUTEST_KERNEL_CODE;
+				ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
+				if (ret < 0) {
+					printf("ERROR: Failed to fetch test address from kernel\n");
+					return ERROR;
```
Comment (thapav) [inline apps/examples/mpu/mpu_test.c:169]: Same as above, Please close tc_fd before return.

File: apps/examples/mpu/mpu_test.c [hunk 11]
```diff
@@ -132,53 +149,64 @@ int mpu_tc_main(int argc, char *argv[])
...
-				*address = dest;
+				*(obj.addr) = dest;
 
 				printf("ERROR: User Task made invalid access to Kernel code space\n");
 			} else if (ch == 'D' || ch == 'd') {
-				address = (uint32_t *)CONFIG_MPUTEST_KERNEL_DATA_ADDR;
+				obj.type = MPUTEST_KERNEL_DATA;
+				ret = ioctl(tc_fd, TESTIOC_MPUTEST, &obj);
+				if (ret < 0) {
+					printf("ERROR: Failed to fetch test address from kernel\n");
```
Comment (thapav) [inline apps/examples/mpu/mpu_test.c:189]: Same as above, Please close tc_fd before return.

File: loadable_apps/loadable_sample/wifiapp/recovery.c [hunk 12]
```diff
@@ -44,18 +51,31 @@ static void *assert_thread(void *index)
...
 	}
 
+	close(tc_fd);
```
Comment (thapav) [inline loadable_apps/loadable_sample/wifiapp/recovery.c:78]: Same as above,
Please close tc_fd before return for all occurences of failure or alternatively you can add it to errout flag.

**Comment by thapav** [comment]
Fix commit title:
os/arch: Fix issue in fault handling**assert address detection

File: apps/examples/mpu/mpu_test.c (hunk 5 above)
Comment (jeongchanKim) [reply apps/examples/mpu/mpu_test.c:36]: We checked that current driver is okay. (I mean we can use open without drv implementation.)
Please check my second comment only.

Comment (kishore-sn) [reply apps/examples/mpu/mpu_test.c]: Done.

File: os/drivers/testcase/kernel_testcase_drv.c [hunk 13]
```diff
@@ -153,5 +204,5 @@ static ssize_t kernel_test_drv_write(FAR struct file *filep, FAR const char *buf
 
 void kernel_test_drv_register(void)
 {
-	(void)register_driver(KERNEL_TC_DRVPATH, &kernel_test_drv_fops, 0666, NULL);
+	(void)register_driver(TESTCASE_DRVPATH, &kernel_test_drv_fops, 0666, NULL);
```
Comment (jeongchanKim) [inline os/drivers/testcase/kernel_testcase_drv.c:207]: Please change the function name and operation variable name also.

File: apps/examples/mpu/Kconfig [hunk 14]
```diff
@@ -6,30 +6,12 @@
 menuconfig EXAMPLES_MPU_TEST
 	bool "MPU TestCase Example"
 	depends on BUILD_PROTECTED
+	select KERNEL_TEST_DRV
```
Comment (jeongchanKim) [inline apps/examples/mpu/Kconfig:9]: Please change the config name, not only for kernel test.

**Comment by sunghan-chang** [comment]
Please check coding rule violations.

File: os/drivers/testcase/kernel_testcase_drv.c (hunk 13 above)
Comment (kishore-sn) [reply os/drivers/testcase/kernel_testcase_drv.c:207]: Done.

**Comment by kishore-sn** [comment]
Added new commit for name change of kernel tc driver.
